crates/topos-tce-api/src/graphql/query.rs

```diff
@@ -102,6 +108,49 @@ impl QueryRoot {
         ctx: &Context<'_>,
         certificate_id: CertificateId,
     ) -> Result<Certificate, GraphQLServerError> {
+        info!("In certificate");
         Self::certificate_by_id(ctx, certificate_id).await
     }
 }
+
+pub struct SubscriptionRoot;
+
+impl SubscriptionRoot {
+    pub(crate) async fn new_transient_stream(
```

Comment:
Docs?

I haven't read the full patch, so perhaps it'll be answered below, but what is a "transient stream" in this concept? What is the role of a `SubscriptionRoot`?

Q: this call looks like it doesn't really need to be async. I guess it's a general async rust question: should we strive to make as much code as possible follow the oneshot channel pattern seen here, or only where it's strictly needed/useful? Like, what is the point of sending the `InternalRuntimeCommand::NewTransientStream` and then wait for it, rather than do the work synchronously? (Honest question, not intended as a critique)

Reply:
`SubscriptionRoot` is a struct used for setting up `GraphQL` schema.
`TransientStream` is described on the struct doc itself.

>  Q: this call looks like it doesn't really need to be async.
How would you do to make it non async then? The ownership of creating the different `Stream` is own by the `topos-tce-api::Runtime`.

> should we strive to make as much code as possible follow the oneshot channel pattern seen here, or only where it's strictly needed/useful?

The `oneshot` pattern is a request/response pattern, and it is the only (or at least the cleanest) to communicate asynchronously with a spawn task. It allows internal back pressure and prevents managing complex future pool and ids.

> Like, what is the point of sending the InternalRuntimeCommand::NewTransientStream and then wait for it, rather than do the work synchronously? (Honest question, not intended as a critique)

On one side we have the GraphQL API that run into its own task, on the other side we have the gRPC runtime that run also in its own task, and in the middle we have the topos API runtime which create the link between the rest of the internal topos component and those 2 frontend API. I'm curious to know how you would implement that without blocking everything.

Reply:
What I had in mind to ask was essentially this code:

```rust
        use std::sync::mpsc::sync_channel;
        let (sender, receiver) = sync_channel(255);
        register.send(InternalRuntimeCommand::NewTransientStream { sender });
        let stream = receiver
            .recv()
            .map_err(|_| GraphQLServerError::InternalError("Bad"))?
            .map_err(|e| GraphQLServerError::TransientStream(e.to_string()))?;
        Ok(stream.map(|c| c.into()))
```

Why would the above not work? It would block (obviously), but we are unable to make progress on setting up the stream anyway so why not simply block until the other side has time to do the work? We `await` twice here, paying the overhead of pausing the task and yield, then awake and resume twice. Are we really creating so many new transient streams that it's worth paying the cost?

There is a cost to complex code and async rust **is** harder than sync rust. Harder to read, teach and debug.

I'm not saying this is bad code. I would like us to stop and think about applying async patterns sometimes. And be careful to apply it only when it's needed.

Reply:
Async rust is pretty straightforward and documented, both in `Tokio` and standard library.
Making things blocking for no reason is a bad design I think, blocking stuff must be used only for critical things  or internal mutation, which is not the case here.

The `sync_channel` blocks the current thread, preventing any future to progress until the tasks finish. Do we want to block an entire thread for each API calls that could create a stream? I don't think so.

> I would like us to stop and think about applying async patterns sometimes. And be careful to apply it only when it's needed.

Don't worry :)
We know where we need to remove async stuff on the codebase. To stabilize the project we split the whole codebase into multiple chunks to have a better separation of concern between domain and logic. We are reducing those components now that we have a better view.

Reply:
I'm glad you're finding it straight forward. :)
Many think otherwise, and regardless of how easy or hard it is for people, there is little doubt that debugging async code is (much) harder than sync code.

> Do we want to block an entire thread for each API calls that could create a stream? I don't think so.

You are correct of course. Did you know that I didn't realise this was being called once for each connection until I read this comment? I.e. that it's part of the setup for each graphql client. When I keep on harping on the need to document public methods this is what I mean: there will be readers of this code that are slow and inexperienced with the system and when we add a line of documentation, even if it's just stating the obvious, we end up saving readers like that some precious time.

Reply:
I mean, the `GraphQL` method that use this method is explaining a lot of things already.
The type `TransientStream` is also explain a lot of things about how it works.
I don't think we need to re-explain everything again for the third time.